storage/src/mmr/journaled.rs

```diff
@@ -0,0 +1,330 @@
+        let mut vec = Vec::new();
+        let oldest_peak = PeakIterator::new(journal_size).next().unwrap().0;
+        for i in oldest_peak..journal_size {
+            let item = journal.read(i).await?;
+            vec.push(H::Digest::try_from(item.as_ref()).unwrap());
+        }
+        let mem_mmr = MemMmr::init(vec, oldest_peak);
+
+        let mut s = Self {
+            mem_mmr,
+            journal,
+            journal_size,
+        };
+
+        if orphaned_leaf != zero_digest {
+            // Recover the orphaned leaf and any missing parents.
+            let mut hasher = H::new();
+            let pos = s.add(&mut hasher, &orphaned_leaf);
+            assert!(pos == journal_size);
+            warn!("recovered orphaned leaf: {:?}", orphaned_leaf);
+            s.sync().await?;
+        }
+        Ok(s)
+    }
+
+    /// Add an element to the MMR and return its position in the MMR. Elements added to the MMR
+    /// aren't persisted to disk until `sync` is called.
+    pub fn add(&mut self, h: &mut H, element: &H::Digest) -> u64 {
+        self.mem_mmr.add(h, element)
+    }
+
+    pub fn root(&mut self, h: &mut H) -> H::Digest {
+        self.mem_mmr.root(h)
+    }
+
+    /// Sync any new elements to disk.
+    pub async fn sync(&mut self) -> Result<(), Error> {
+        for i in self.journal_size..self.mem_mmr.size() {
+            let node = self.mem_mmr.get_node(i).await?.unwrap();
+            self.journal
+                .append(node.as_ref().try_into().unwrap())
+                .await?;
+        }
+        self.journal_size = self.mem_mmr.size();
+        self.mem_mmr.forget_max();
+        self.journal.sync().await?;
+        Ok(())
+    }
+
+    /// Close the journal
+    pub async fn close(mut self) -> Result<(), Error> {
+        self.sync().await?;
+        self.journal.close().await.map_err(Error::JournalError)
+    }
+
+    pub async fn proof(&self, element_pos: u64) -> Result<Proof<H>, Error> {
+        self.range_proof(element_pos, element_pos).await
+    }
+
+    /// Return an inclusion proof for the specified range of elements, inclusive of both endpoints.
+    ///
+    /// Returns ElementPruned error if some element needed to generate the proof has been pruned.
+    pub async fn range_proof(
+        &self,
+        start_element_pos: u64,
+        end_element_pos: u64,
+    ) -> Result<Proof<H>, Error> {
+        Proof::<H>::range_proof::<Mmr<B, E, H, DIGEST_LENGTH>>(
+            self,
+            start_element_pos,
+            end_element_pos,
+        )
+        .await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{Blob, JConfig, Mmr, RStorage, Storage};
+    use commonware_cryptography::{sha256::Digest, Hasher, Sha256};
+    use commonware_macros::test_traced;
+    use commonware_runtime::{deterministic::Executor, Runner};
+    use prometheus_client::registry::Registry;
+    use std::sync::{Arc, Mutex};
+
+    fn test_digest(v: u8) -> Digest {
```

Comment:
You can just do `hash(&v.to_be_bytes())` inside this function

Reply:
done